// @nextui-org/popover/dist/chunk-JBPODFPT.mjs
"use client";

// src/use-aria-popover.ts
import { useLayoutEffect } from "react";
import { useOverlay, useOverlayPosition } from "@react-aria/overlays";
import { toReactAriaPlacement, ariaHideOutside } from "@nextui-org/aria-utils";
import { mergeProps } from "@react-aria/utils";
function useReactAriaPopover(props, state) {
  const {
    triggerRef,
    popoverRef,
    showArrow,
    offset = 7,
    crossOffset = 0,
    scrollRef,
    shouldFlip,
    placement: placementProp = "top",
    containerPadding,
    isNonModal: isNonModalProp,
    isKeyboardDismissDisabled,
    ...otherProps
  } = props;
  const isNonModal = isNonModalProp || true;
  const { overlayProps, underlayProps } = useOverlay(
    {
      isOpen: state.isOpen,
      onClose: state.close,
      shouldCloseOnBlur: true,
      isDismissable: !isNonModal,
      isKeyboardDismissDisabled,
      shouldCloseOnInteractOutside(element) {
        let trigger = triggerRef == null ? void 0 : triggerRef.current;
        return !trigger || !trigger.contains(element);
      }
    },
    popoverRef
  );
  const {
    overlayProps: positionProps,
    arrowProps,
    placement
  } = useOverlayPosition({
    ...otherProps,
    scrollRef,
    crossOffset,
    shouldFlip,
    containerPadding,
    targetRef: triggerRef,
    overlayRef: popoverRef,
    isOpen: state.isOpen,
    placement: toReactAriaPlacement(placementProp),
    offset: showArrow ? offset + 3 : offset,
    onClose: () => {
    }
  });
  useLayoutEffect(() => {
    if (state.isOpen && !isNonModal && popoverRef.current) {
      return ariaHideOutside([popoverRef.current]);
    }
  }, [isNonModal, state.isOpen, popoverRef]);
  return {
    popoverProps: mergeProps(overlayProps, positionProps),
    arrowProps,
    underlayProps,
    placement
  };
}

export {
  useReactAriaPopover
};
