// @nextui-org/user/dist/chunk-PLYHOYRQ.mjs
"use client";

// src/use-user.ts
import { useMemo, useCallback } from "react";
import { useFocusRing } from "@react-aria/focus";
import { user } from "@nextui-org/theme";
import { clsx, dataAttr } from "@nextui-org/shared-utils";
import { useDOMRef } from "@nextui-org/react-utils";
import { mergeProps } from "@react-aria/utils";
function useUser(props) {
  const {
    as,
    ref,
    name,
    description,
    className,
    classNames,
    isFocusable = false,
    avatarProps = {
      isFocusable: false,
      name: typeof name === "string" ? name : void 0
    },
    ...otherProps
  } = props;
  const Component = as || "div";
  const domRef = useDOMRef(ref);
  const { isFocusVisible, isFocused, focusProps } = useFocusRing({});
  const canBeFocused = useMemo(() => {
    return isFocusable || as === "button";
  }, [isFocusable, as]);
  const slots = useMemo(() => user(), []);
  const baseStyles = clsx(classNames == null ? void 0 : classNames.base, className);
  const getUserProps = useCallback(
    () => ({
      ref: domRef,
      tabIndex: canBeFocused ? 0 : -1,
      "data-focus-visible": dataAttr(isFocusVisible),
      "data-focus": dataAttr(isFocused),
      className: slots.base({
        class: baseStyles
      }),
      ...mergeProps(otherProps, canBeFocused ? focusProps : {})
    }),
    [canBeFocused, slots, baseStyles, focusProps, otherProps]
  );
  return {
    Component,
    className,
    slots,
    name,
    description,
    classNames,
    baseStyles,
    focusProps,
    avatarProps,
    getUserProps
  };
}

export {
  useUser
};
